### PR TITLE
[Fix][Engram] Fix inter-block global memory write-read data race in `EngramGateConvBwdKernel`

### LINKED ISSUE #1729: [Bug] EngramGateConvBwdKernel: inter-block global memory read-write race
The Pass 1b launch of `EngramGateConvBwdKernel`([`tileops/kernels/engram/engram_bwd.py` L148-L210](https://github.com/tile-ai/TileOPs/blob/48782d50e698c00d053e8cd0b2f2625e6ca54bdd/tileops/kernels/engram/engram_bwd.py#L148-L210)) both **reads** and **writes** the same global buffer `dvhat_buf` in one launch. Each block handles one sequence position `t`: it **reads** a conv-transpose buffer  `dvhat_buf` rows `t+1, t+2, t+3` of *other* blocks — and **writes** its own row `t`.

This is an **inter-block, global-memory write-read race**: block `t` reads row `t+1` while block `t+1` overwrites it, so the gradients `dH` / `dk` / `dv` come out **non-deterministic** and **wrong**. 

### Environment
TileOPs `48782d5` (current main)
Python: 3.10.19 (Linux, x86_64) 
TileLang: 0.1.12 
torch: 2.10.0+cu128

GPU: NVIDIA GeForce RTX 5070 Ti (sm_120)
CUDA: 12.8 

### Source Code in TileLang DSL (buggy)

Pass 1b in [`tileops/kernels/engram/engram_bwd.py` L148-L210](https://github.com/tile-ai/TileOPs/blob/48782d50e698c00d053e8cd0b2f2625e6ca54bdd/tileops/kernels/engram/engram_bwd.py#L148-L210):

```python
with T.Kernel(seq_len, M, threads=threads) as (bx, by):
    bid, tid = by, bx                                     
    ...
    for p in T.serial(KS):                                 
        dst_t = tid + (KS - 1) - p                         # dst_t: {tid, tid+1, tid+2, tid+3}
        for j in T.Parallel(d_padded):
            d_si = T.if_then_else((dst_t >= 0) * (dst_t < seq_len),
                                  dvhat_buf[bid, dst_t, j], 0.0)   # L169  READ dvhat_buf rows (tid..tid+3)
    ...
    for j in T.Parallel(d_padded):
        dvhat_buf[bid, tid, j] = dvhat_local[j]            # L210  WRITE dvhat_buf row (tid)
```
### Compiled TVM TIR

Both racing accesses land in `main_kernel_3` (the Pass 1b launch), with no barrier between them. At the dump config `seq_len=64`: each block handles one sequence position, and `d_padded=256` is split over 128 threads (2 each):

```python
tid = T.launch_thread("blockIdx.x", 64)          # one block per sequence position
...
# conv-transpose window read (row dst_t = tid+3-p):
dvhat_buf_local[0:2] = T.if_then_else(tid - p < 61,
    dvhat_buf_1[tid*256 + tx*2 + 768 - p*256 : ... + 2], 0)      # <-- READ
...
# overwrite own row tid, in place:
dvhat_buf_1[tid*256 + tx*2 : tid*256 + tx*2 + 2] = dvhat_local_1[0:2]   # <-- WRITE
```

Concretely: block `tid` at `p=2` reads address `(tid+1)*256 + tx*2`; block `tid+1` writes `(tid+1)*256 + tx*2`. Same global address, different blocks (`blockIdx.x = tid` vs `tid+1`), same launch, no barrier — a write-read race.

### Reproduction

Uses TileOPs' own test infrastructure. Run from the TileOPs repo root.

```bash
git clone https://github.com/tile-ai/TileOPs.git
cd TileOPs && git checkout 48782d50e698c00d053e8cd0b2f2625e6ca54bdd
pip install -e .
pip install tilelang==0.1.12
# sm_120 only: export CUDA_HOME=/usr/local/cuda-13.1 PATH=/usr/local/cuda-13.1/bin:$PATH
PYTHONPATH=. python reproduce.py            # from the repo root
```

reproduce.py:
```python
import torch
from workloads.engram import EngramGateConvBwdTest
from tileops.ops.engram import EngramGateConvBwdOp
from tests.ops.test_engram import ref_engram_gate_conv_bwd
from tileops.kernels.engram import EngramGateConvBwdKernel

# sm_120 architecture
EngramGateConvBwdKernel.supported_archs += [120]

torch.manual_seed(0)
# grid = seq_len * M blocks; seq_len=512, M=2 -> 1024 blocks (oversubscribe the ~70 SMs)
M, seq_len, d = 2, 512, 256
inp = EngramGateConvBwdTest(M, seq_len, d, torch.float16, 1e-6).gen_inputs()
op = EngramGateConvBwdOp(M, seq_len, d, torch.float16)
dH_r, dk_r, dv_r, *_ = ref_engram_gate_conv_bwd(*inp, 1e-6)

# run 100 times to check result determinism
base = [o.clone() for o in op.forward(*inp)[:3]]
nondet = [0, 0, 0]; dv_err = 0.0
for _ in range(100):
    dH, dk, dv, *_ = [o.clone() for o in op.forward(*inp)]
    for i, g in enumerate((dH, dk, dv)):
        nondet[i] += not torch.equal(g, base[i])
    dv_err = max(dv_err, (dv.float() - dv_r.float()).abs().max().item())

print(f"dH/dk/dv nondeterministic: {nondet}/100   (race-free => [0, 0, 0])")
print(f"dv max err vs reference:   {dv_err:.4f}    (test uses atol 0.2)")
```

### Expected behavior

A race-free kernel is deterministic:
```
dH/dk/dv nondeterministic: [0, 0, 0]/100   (race-free => [0, 0, 0])
dv max err vs reference:   0.0002    (test uses atol 0.2)
```

Actual output on `sm_120` at `seq_len=512` (the `dv` error magnitude varies run-to-run because it is a race):
```
dH/dk/dv nondeterministic: [100, 100, 100]/100   (race-free => [0, 0, 0])
dv max err vs reference:   0.5263    (test uses atol 0.2)
```

The race fires on every run and corrupts the gradients: `dv`'s max error against the reference is `0.5`, versus `0.0002` on a race-free run.

### Suggested Fix

Give Pass 1b its own output buffer `dvhat_out` so `dvhat_buf` stays **read-only** within the launch: Pass 1b reads the window from `dvhat_buf` and writes the result to `dvhat_out`; the next launch (Pass 2) reads `dvhat_out`. The two essential changes:

```diff
@@ -209,2 +209,2 @@ Pass 1b: store total d(vhat)
                 for j in T.Parallel(d_padded):
-                    dvhat_buf[bid, tid, j] = dvhat_local[j]
+                    dvhat_out[bid, tid, j] = dvhat_local[j]
@@ -235,2 +235,2 @@ Pass 2: load d(vhat)
                 for j in T.Parallel(d_padded):
-                    dvhat_local[j] = dvhat_buf[bid, tid, j]
+                    dvhat_local[j] = dvhat_out[bid, tid, j]
```
plus registering `dvhat_out` as a new output buffer of the JIT kernel.

### Other Notes

`test_engram_gate_conv_bwd`'s configs top out at **128 blocks** (grid = `seq_len × M`), too few to surface the race — `0/60` across all of them. It takes many more: `seq_len=512` (1024 blocks) can reproduce the race reliably, and real sequence lengths give thousands. I suggest to add a large-`seq_len` config and assert determinism on `dH`/`dk`/`dv`:

```diff
@@ -150,1 +150,2 @@ class EngramGateConvBwdFixture(FixtureBase):
             pytest.param(2, 16, 256, torch.bfloat16, False, marks=pytest.mark.full),
+            pytest.param(2, 512, 256, torch.float16, False, marks=pytest.mark.full),
@@ -162,1 +162,4 @@ def test_engram_gate_conv_bwd(M, seq_len, d, dtype, tune):
     test.check(op, *inputs, atol=atol, rtol=rtol)
+    a, b = op(*inputs), op(*inputs)
+    for i, name in ((0, "dH"), (1, "dk"), (2, "dv")):
+        assert torch.equal(a[i], b[i]), f"{name} non-deterministic across runs (data race)"
```

### PR DESCRIPTION
**Pull request by arxgy**
In [`EngramGateConvBwdKernel`](https://github.com/tile-ai/TileOPs/blob/48782d50e698c00d053e8cd0b2f2625e6ca54bdd/tileops/kernels/engram/engram_bwd.py#L148-L210)'s Pass 1b, each block reads its neighbours' rows of `dvhat_buf` while overwriting its own, with no cross-block barrier between them — an inter-block global-memory write-read race. Once the grid (`seq_len * M` blocks) oversubscribes the SMs, `dH` / `dk` / `dv` come out non-deterministic and wrong.

This PR fixes this race by storing `d(vhat)` to a separate `dvhat_out` buffer (read back in Pass 2), leaving `dvhat_buf` read-only within Pass 1b. The computed values and other kernel behaviors are unchanged, with no measured perf change.

### Bug Fix

- Fixes #1729

### Regression Test

Adds a `seq_len=512` (1024-block) config that exposes the race and a determinism assertion (a second run must match `dH` / `dk` / `dv`; `allclose` is masked by its `rtol` tolerance). Fails on the current kernel, passes on the fix. `scripts/test_node_delta.py`: 13 -> 14 (+1).

### CODE REVIEW

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
